optax/contrib/_sophia_h.py

```diff
@@ -0,0 +1,307 @@
+  the hessian diagonal. It must accept `params` as its only argument and return
+  only a scalar (the loss).
+
+  For example, considering loss function `loss_fn(params, batch) -> loss, aux`
+  that takes multiple arguments and returns multiple outputs, we must modify it
+  to `loss_fn(params) -> loss` like below:
+
+  `sophia_obj_fn = lambda params: loss_fn(params, batch)[0]`
+
+  Then it can be passed to sophia's update function:
+
+  `updates, state = sophia.update(updates, state, params, obj_fn=sophia_obj_fn)`
+
+  Notes:
+    - TODO filter non-differentiable inputs to jvp
+    - Paper uses gaussians for hutchinson's estimator but we use rademacher, see
+      https://www.ethanepperly.com/index.php/2024/01/28/dont-use-gaussians-in-stochastic-trace-estimation/
+      and https://x.com/dlwh/status/1785068009016672566
+    - If using `jax.pmap`, providing the pmap axis name will perform separate
+      monte carlo samples on each device for hutchinson's estimator for (almost)
+      free, theoretically increasing hessian estimation accuracy.
+
+  References:
+    Liu et al., `Sophia: A Scalable Stochastic Second-order Optimizer for
+    Language Model Pre-training <https://arxiv.org/abs/2305.14342>`_, 2023
+
+    `Levanter <https://www.github.com/stanford-crfm/levanter>`_
+
+  Args:
+    b1: Exponential decay rate for the first moment estimates.
+    b2: Exponential decay rate for the hessian diagonal estimates. Keep in mind
+        effective `b2` is `1 - (1 - b2) / update_interval`, e.g. default `b2`
+        of 0.99 is effectively 0.999 because default `update_interval` is every
+        10.
+    eps: Small constant to avoid division by zero.
+    gamma: Normalizing constant for the hessian diagonal.
+    clip_threshold: Threshold for clipping updates.
+    update_interval: Interval for updating the hessian diagonal.
+    mu_dtype: dtype of the first moment estimates.
+    pmap_axis_name: Provide pmap axis name if using pmap to perform separate
+        monte carlo samples on each device for hutchinson's estimator for
+        (almost) free.
+    seed: PRNGKey.
+    print_win_rate_every_n_steps: Print sophia win rate every n steps for
+        diagnostic purposes. Authors state this value should stay between
+        0.1 and 0.5 during training. If win rate is too low, try increasing
+        `gamma`.
+
+  Returns:
+    optax.GradientTransformationExtraArgs
+  """
+  tx = [
+      scale_by_sophia_h(
+          b1=b1,
+          b2=b2,
+          eps=eps,
+          gamma=gamma,
+          clip_threshold=clip_threshold,
+          update_interval=update_interval,
+          mu_dtype=mu_dtype,
+          pmap_axis_name=pmap_axis_name,
+          seed=seed,
+          print_win_rate_every_n_steps=print_win_rate_every_n_steps,
+      ),
+      transform.add_decayed_weights(weight_decay, mask=mask),
+      transform.scale_by_learning_rate(learning_rate),
+  ]
+  return chain(*tx)
+
+
+def _tree_rademacher_like(key, tree):
```

Comment:
Could you use `optax.tree_utils.tree_random_like`?
If the API of `optax.tree_utils.tree_random_like` does not match your needs, just keep what you have but feedback on `optax.tree_utils.tree_random_like` would be useful for us.

Reply:
Ah I didn't know this existed, definitely